File: frontend/src/types/gamification.ts
/**
 * Gamification Types and Interfaces
 * Defines the structure for user progression, badges, achievements, and unlocks
 */

export interface UserProgress {
  user_id: string;
  level: number;
  total_xp: number;
  current_streak: number;
  longest_streak: number;
  skill_points: Record<string, number>;
  unlocked_features: string[];
  last_active_date: string;
  level_progress: {
    current_level_xp: number;
    next_level_xp: number;
    xp_to_next_level: number;
    xp_progress: number;
    percentage: number;
  };
  recent_badges: Badge[];
  recent_xp: XPTransaction[];
}

export interface Badge {
  id: string;
  badge_id: string;
  name: string;
  description: string;
  icon: string;
  category: 'completion' | 'skill' | 'social' | 'special';
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  earned_date: string;
  is_featured: boolean;
  xp_reward?: number;
}

export interface XPTransaction {
  amount: number;
  source: string;
  description: string;
  created_at: string;
  metadata?: Record<string, any>;
}

export const XP_REWARDS = {
  PROJECT_COMPLETE: 100,
  STREAK_DAY: 25,
  STREAK_WEEK: 200,
  COLLABORATION_JOIN: 150,
  HELP_PEER: 50,
  CODE_QUALITY_HIGH: 75,
  INDUSTRY_PROJECT: 250,
  FIRST_PROJECT: 50,
  WEEKLY_STREAK: 100,
  MONTHLY_STREAK: 500,
  COLLABORATION_MASTER: 200,
  MENTOR_HELP: 75,
  BUG_FIX: 25,
  FEATURE_ADD: 50,
  CODE_REVIEW: 30,
  DOCUMENTATION: 40,
  TESTING: 35,
  REFACTORING: 45,
  LEARNING_NEW_TECH: 60,
  COMMUNITY_CONTRIBUTION: 80
} as const;

export interface StreakData {
  current_streak: number;
  longest_streak: number;
  streak_updated: boolean;
  xp_earned: number;
}

export interface LeaderboardEntry {
  rank: number;
  username: string;
  score: number;
  level: number;
  total_xp: number;
  current_streak: number;
}

export interface Leaderboard {
  period: 'weekly' | 'monthly' | 'all_time';
  category: 'xp' | 'streak' | 'projects' | 'collaboration';
  leaderboard: LeaderboardEntry[];
  generated_at: string;
}

export interface UnlockableFeature {
  name: string;
  display_name: string;
  description: string;
  icon: string;
  category: string;
  requirements: {
    level?: number;
    projects_completed?: number;
    total_xp?: number;
    streak_days?: number;
  };
}

export interface UserUnlocks {
  unlocked_features: string[];
  available_unlocks: UnlockableFeature[];
}

export interface SkillTree {
  id: string;
  name: string;
  description: string;
  icon: string;
  color: string;
  is_active: boolean;
  nodes: SkillNode[];
}

export interface SkillNode {
  id: string;
  skill_tree_id: string;
  name: string;
  description: string;
  icon: string;
  level: number;
  xp_cost: number;
  requirements: {
    prerequisite_nodes?: string[];
    min_level?: number;
  };
  benefits: {
    xp_multiplier?: number;
    unlock_feature?: string;
  };
  position_x: number;
  position_y: number;
  is_active: boolean;
  is_unlocked: boolean;
}

export interface Achievement {
  id: string;
  name: string;
  description: string;
  icon: string;
  category: string;
  rarity: 'common' | 'rare' | 'epic' | 'legendary';
  xp_reward: number;
  requirements: Record<string, any>;
  is_hidden: boolean;
  progress: number;
  is_earned: boolean;
}

export interface LevelUpData {
  old_level: number;
  new_level: number;
  xp_gained: number;
  new_unlocks: string[];
  new_badges: Badge[];
  rewards: {
    xp_bonus: number;
    skill_points: number;
    features_unlocked: string[];
  };
}

export interface GamificationStats {
  total_users: number;
  total_xp_awarded: number;
  total_badges_earned: number;
  average_level: number;
  top_performers: LeaderboardEntry[];
  recent_activity: {
    user: string;
    action: string;
    xp_earned: number;
    timestamp: string;
  }[];
}

export interface XPRequest {
  source: string;
  amount: number;
  description: string;
  metadata?: Record<string, any>;
}

export interface StreakRequest {
  activity_type: string;
}

export interface BadgeFeatureRequest {
  badge_id: string;
  is_featured: boolean;
}

// XP Sources
export const XP_SOURCES = {
  PROJECT_COMPLETE: 'project_complete',
  PROJECT_COMPLETE_BEGINNER: 'project_complete_beginner',
  PROJECT_COMPLETE_INTERMEDIATE: 'project_complete_intermediate',
  PROJECT_COMPLETE_ADVANCED: 'project_complete_advanced',
  STREAK_DAY: 'streak',
  STREAK_WEEK: 'streak_milestone',
  STREAK_MONTH: 'streak_milestone',
  COLLABORATION_JOIN: 'collaboration_join',
  COLLABORATION_COMPLETE: 'collaboration_complete',
  HELP_PEER: 'help_peer',
  CODE_QUALITY_HIGH: 'code_quality_high',
  CODE_QUALITY_PERFECT: 'code_quality_perfect',
  INDUSTRY_PROJECT: 'industry_project',
  FIRST_PROJECT: 'first_project',
  DAILY_LOGIN: 'daily_login',
  WEEKLY_GOAL: 'weekly_goal',
  MONTHLY_GOAL: 'monthly_goal',
  SKILL_NODE_UNLOCK: 'skill_node_unlock',
  BADGE_EARNED: 'badge_earned',
  ACHIEVEMENT_EARNED: 'achievement_earned',
  LEADERBOARD_TOP_10: 'leaderboard_top_10',
  LEADERBOARD_TOP_3: 'leaderboard_top_3',
  LEADERBOARD_FIRST: 'leaderboard_first',
  LEVEL_UP: 'level_up'
} as const;

// Badge Categories
export const BADGE_CATEGORIES = {
  COMPLETION: 'completion',
  SKILL: 'skill',
  SOCIAL: 'social',
  SPECIAL: 'special'
} as const;

// Rarity Levels
export const RARITY_LEVELS = {
  COMMON: 'common',
  RARE: 'rare',
  EPIC: 'epic',
  LEGENDARY: 'legendary'
} as const;

// Leaderboard Periods
export const LEADERBOARD_PERIODS = {
  WEEKLY: 'weekly',
  MONTHLY: 'monthly',
  ALL_TIME: 'all_time'
} as const;

// Leaderboard Categories
export const LEADERBOARD_CATEGORIES = {
  XP: 'xp',
  STREAK: 'streak',
  PROJECTS: 'projects',
  COLLABORATION: 'collaboration'
} as const;

// Feature Unlocks
export const FEATURE_UNLOCKS = {
  COLLABORATION: 'collaboration',
  MENTORING: 'mentoring',
  INDUSTRY_CHALLENGES: 'industry_challenges',
  PORTFOLIO_CERTIFICATION: 'portfolio_certification',
  ADVANCED_EDITOR: 'advanced_editor',
  TEAM_FORMATION: 'team_formation',
  CUSTOM_PROJECTS: 'custom_projects',
  AI_ASSISTANT_ADVANCED: 'ai_assistant_advanced'
} as const;

// Skill Tree Categories
export const SKILL_TREE_CATEGORIES = {
  FRONTEND: 'frontend',
  BACKEND: 'backend',
  DATA_SCIENCE: 'data_science',
  AI_ML: 'ai_ml',
  MOBILE: 'mobile',
  DEVOPS: 'devops',
  GAME_DEV: 'game_dev',
  BLOCKCHAIN: 'blockchain'
} as const;

// Achievement Categories
export const ACHIEVEMENT_CATEGORIES = {
  PROJECTS: 'projects',
  STREAKS: 'streaks',
  COLLABORATION: 'collaboration',
  SKILLS: 'skills',
  SOCIAL: 'social',
  SPECIAL: 'special',
  MILESTONE: 'milestone'
} as const;

export default UserProgress;

File: backend/app/models/gamification.py
"""
Gamification Models
Database models for user progression, badges, achievements, and unlocks
"""

from sqlalchemy import Column, Integer, String, DateTime, Boolean, Float, Text, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid

from ..core.database import Base

class UserProgress(Base):
    __tablename__ = "user_progress"
    __table_args__ = {'extend_existing': True}
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String, ForeignKey("users.id"), unique=True, nullable=False)
    level = Column(Integer, default=1, nullable=False)
    total_xp = Column(Integer, default=0, nullable=False)
    current_streak = Column(Integer, default=0, nullable=False)
    longest_streak = Column(Integer, default=0, nullable=False)
    skill_points = Column(JSON, default=dict)  # {"python": 100, "javascript": 50, ...}
    unlocked_features = Column(JSON, default=list)  # ["collaboration", "mentoring", ...]
    last_active_date = Column(DateTime, default=datetime.utcnow)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    user = relationship("User", back_populates="progress")
    badges = relationship("UserBadge", back_populates="user_progress")
    xp_transactions = relationship("XPTransaction", back_populates="user_progress")
    streak_history = relationship("StreakHistory", back_populates="user_progress")

class Badge(Base):
    __tablename__ = "badges"
    __table_args__ = {'extend_existing': True}
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=False)
    icon = Column(String(50), nullable=False)  # Icon identifier
    category = Column(String(20), nullable=False)  # completion, skill, social, special
    rarity = Column(String(20), default="common")  # common, rare, epic, legendary
    xp_reward = Column(Integer, default=0)
    requirements = Column(JSON, default=dict)  # {"projects_completed": 5, "streak_days": 7}
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    user_badges = relationship("UserBadge", back_populates="badge")

class UserBadge(Base):
    __tablename__ = "user_badges"
    __table_args__ = {'extend_existing': True}
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_progress_id = Column(String, ForeignKey("user_progress.id"), nullable=False)
    badge_id = Column(String, ForeignKey("badges.id"), nullable=False)
    earned_date = Column(DateTime, default=datetime.utcnow)
    is_featured = Column(Boolean, default=False)  # Show on profile
    
    # Relationships
    user_progress = relationship("UserProgress", back_populates="badges")
    badge = relationship("Badge", back_populates="user_badges")

class XPTransaction(Base):
    __tablename__ = "xp_transactions"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_progress_id = Column(String, ForeignKey("user_progress.id"), nullable=False)
    amount = Column(Integer, nullable=False)
    source = Column(String(50), nullable=False)  # project_complete, streak, collaboration, etc.
    description = Column(String(200), nullable=False)
    transaction_metadata = Column(JSON, default=dict)  # Additional context
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    user_progress = relationship("UserProgress", back_populates="xp_transactions")

class StreakHistory(Base):
    __tablename__ = "streak_history"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_progress_id = Column(String, ForeignKey("user_progress.id"), nullable=False)
    date = Column(DateTime, nullable=False)
    activity_type = Column(String(50), nullable=False)  # coding, project_complete, login
    xp_earned = Column(Integer, default=0)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    user_progress = relationship("UserProgress", back_populates="streak_history")

class UnlockableFeature(Base):
    __tablename__ = "unlockable_features"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(50), unique=True, nullable=False)
    display_name = Column(String(100), nullable=False)
    description = Column(Text, nullable=False)
    icon = Column(String(50), nullable=False)
    requirements = Column(JSON, nullable=False)  # {"level": 3, "projects_completed": 5}
    category = Column(String(30), nullable=False)  # collaboration, mentoring, challenges, etc.
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)

class Leaderboard(Base):
    __tablename__ = "leaderboards"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    period = Column(String(20), nullable=False)  # weekly, monthly, all_time
    category = Column(String(30), nullable=False)  # xp, streak, projects, collaboration
    rank = Column(Integer, nullable=False)
    score = Column(Integer, nullable=False)
    period_start = Column(DateTime, nullable=False)
    period_end = Column(DateTime, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    user = relationship("User")

class SkillTree(Base):
    __tablename__ = "skill_trees"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(50), nullable=False)
    description = Column(Text, nullable=False)
    icon = Column(String(50), nullable=False)
    color = Column(String(7), nullable=False)  # Hex color
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)

class SkillNode(Base):
    __tablename__ = "skill_nodes"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    skill_tree_id = Column(String, ForeignKey("skill_trees.id"), nullable=False)
    name = Column(String(50), nullable=False)
    description = Column(Text, nullable=False)
    icon = Column(String(50), nullable=False)
    level = Column(Integer, nullable=False)  # 1-5
    xp_cost = Column(Integer, nullable=False)
    requirements = Column(JSON, default=dict)  # {"prerequisite_nodes": [], "min_level": 1}
    benefits = Column(JSON, default=dict)  # {"xp_multiplier": 1.1, "unlock_feature": "advanced_editor"}
    position_x = Column(Float, nullable=False)  # For UI positioning
    position_y = Column(Float, nullable=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    skill_tree = relationship("SkillTree")

class UserSkillNode(Base):
    __tablename__ = "user_skill_nodes"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_progress_id = Column(String, ForeignKey("user_progress.id"), nullable=False)
    skill_node_id = Column(String, ForeignKey("skill_nodes.id"), nullable=False)
    unlocked_date = Column(DateTime, default=datetime.utcnow)
    
    # Relationships
    user_progress = relationship("UserProgress")
    skill_node = relationship("SkillNode")

class Achievement(Base):
    __tablename__ = "achievements"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=False)
    icon = Column(String(50), nullable=False)
    category = Column(String(30), nullable=False)
    rarity = Column(String(20), default="common")
    xp_reward = Column(Integer, default=0)
    requirements = Column(JSON, nullable=False)
    is_hidden = Column(Boolean, default=False)  # Secret achievements
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)

class UserAchievement(Base):
    __tablename__ = "user_achievements"
    
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_progress_id = Column(String, ForeignKey("user_progress.id"), nullable=False)
    achievement_id = Column(String, ForeignKey("achievements.id"), nullable=False)
    earned_date = Column(DateTime, default=datetime.utcnow)
    progress = Column(Integer, default=0)  # For progress tracking
    
    # Relationships
    user_progress = relationship("UserProgress")
    achievement = relationship("Achievement")

# XP Rewards Configuration
XP_REWARDS = {
    "PROJECT_COMPLETE": 100,
    "PROJECT_COMPLETE_BEGINNER": 75,
    "PROJECT_COMPLETE_INTERMEDIATE": 125,
    "PROJECT_COMPLETE_ADVANCED": 200,
    "STREAK_DAY": 25,
    "STREAK_WEEK": 200,
    "STREAK_MONTH": 1000,
    "COLLABORATION_JOIN": 150,
    "COLLABORATION_COMPLETE": 300,
    "HELP_PEER": 50,
    "CODE_QUALITY_HIGH": 75,
    "CODE_QUALITY_PERFECT": 150,
    "INDUSTRY_PROJECT": 250,
    "FIRST_PROJECT": 200,
    "DAILY_LOGIN": 10,
    "WEEKLY_GOAL": 500,
    "MONTHLY_GOAL": 2000,
    "SKILL_NODE_UNLOCK": 50,
    "BADGE_EARNED": 25,
    "ACHIEVEMENT_EARNED": 100,
    "LEADERBOARD_TOP_10": 100,
    "LEADERBOARD_TOP_3": 250,
    "LEADERBOARD_FIRST": 500
}

# Level Requirements (XP needed for each level)
LEVEL_REQUIREMENTS = {
    1: 0,
    2: 100,
    3: 250,
    4: 450,
    5: 700,
    6: 1000,
    7: 1350,
    8: 1750,
    9: 2200,
    10: 2700,
    11: 3250,
    12: 3850,
    13: 4500,
    14: 5200,
    15: 5950,
    16: 6750,
    17: 7600,
    18: 8500,
    19: 9450,
    20: 10450,
    21: 11500,
    22: 12600,
    23: 13750,
    24: 14950,
    25: 16200,
    26: 17500,
    27: 18850,
    28: 20250,
    29: 21700,
    30: 23200,
    31: 24750,
    32: 26350,
    33: 28000,
    34: 29700,
    35: 31450,
    36: 33250,
    37: 35100,
    38: 37000,
    39: 38950,
    40: 40950,
    41: 43000,
    42: 45100,
    43: 47250,
    44: 49450,
    45: 51700,
    46: 54000,
    47: 56350,
    48: 58750,
    49: 61200,
    50: 63700
}

# Feature Unlock Requirements
FEATURE_UNLOCKS = {
    "collaboration": {"level": 3, "projects_completed": 5},
    "mentoring": {"level": 10, "projects_completed": 15},
    "industry_challenges": {"level": 15, "projects_completed": 25},
    "portfolio_certification": {"level": 25, "projects_completed": 50},
    "advanced_editor": {"level": 5, "projects_completed": 10},
    "team_formation": {"level": 8, "projects_completed": 12},
    "custom_projects": {"level": 12, "projects_completed": 20},
    "ai_assistant_advanced": {"level": 18, "projects_completed": 30}
}
